go-controller/pkg/clustermanager/secondary_network_cluster_controller.go

```diff
@@ -0,0 +1,274 @@
+package clustermanager
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	cache "k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/retry"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+// secondaryNetworkClusterController is the object for managing the secondary
+// layer3 networks for all the nodes. Listens to the node events.
+type secondaryNetworkClusterController struct {
+	networkClusterControllerBase
+	nodeL3NetworkController *NodeNetworkController
+
+	// per controller NAD/netconf name information
+	util.NetInfo
+	util.NetConfInfo
+
+	// Record new nodes we failed to configure
+	addNodeFailed sync.Map
```

Comment:
In the next PR (zone_support - https://github.com/ovn-org/ovn-kubernetes/pull/3169/files#diff-3daaa69dff2f5bb268105a5495c5bb16009d76306db9620e601f53f46535e154),  default network cluster controller along with allocating subnets for each node it also does
     - allocating an id for each node
     - allocating a join subnet for each zone
     -  transit switch ip for each node
     
That's why I created secondaryNetworkClusterController for secondary network controller.

Let me know if you think its better to have have just a single NetworkClusterController in this PR and the next PR can probably create a separate controller or maybe something else depending on the review comments/suggestions from you and others.

Reply:
That sounds like stuff that needs to happen once but that every network might depend on, not only the default network.

Would secondaryNetworkClusterControllers depend (and have to wait for) anything of that? If so, perhaps it could make more sense to have a some other specific (zone) controller to do that and then having a common network controller implementation that would include the wait for that to happen.

So the thing is we might end up adding global things to the default network controller. That's not a problem per-se because there is always a default network controller. But we might miss different ways to organize the code better.

Otherwise, ignore. We can wait how things look with the other PR.

Reply:
Still wondering about the difference with `addNodeFailed` though.

Reply:
@msherif1234 added this in the initial versions.   I think we can remove this addNodeFailed from secondary controller.

Reply:
Since it is there for both for the default and secondary network controllers, I am inclined to think it needs to stay for both the default & secondary cluster network controller. But I am unsure right now what specific corner cases is this suposed to handle, so I will let @msherif1234 comment on that.

Reply:
BTW @numansiddique could you comment on

> Would secondaryNetworkClusterControllers depend (and have to wait for) anything of that? If so, perhaps it could make more sense to have a some other specific (zone) controller to do that and then having a common network controller implementation that would include the wait for that to happen.

Reply:
No.  I don't think secondaryNetworkClusterControllers depends or have to wait for the default cluster controller to do or set something.

I like your approach of having one network controller code which will allocate the subnets.  One instance of it for the default network and an instance for each secondary layer3 network.

And the next zone support PR can add another controller for zone which would do node id allocation, zone join subnet, transit switch ip etc.

I'll explore more on this and see if I can do it in the next push.

Reply:
We can defer this to a later PR if you feel this risks delaying this PR substantially.

Reply:
Done.  The latest push has these changes.  Lot of redundant code is gone now.